prover/witness_generator/src/recursion_tip.rs

```diff
@@ -0,0 +1,341 @@
+            )
+            .await;
+
+        transaction
+            .fri_witness_generator_dal()
+            .mark_recursion_tip_job_as_successful(job_id, started_at.elapsed())
+            .await;
+
+        transaction.commit().await.unwrap();
+        Ok(())
+    }
+
+    fn max_attempts(&self) -> u32 {
+        self.config.max_attempts
+    }
+
+    async fn get_job_attempts(&self, job_id: &L1BatchNumber) -> anyhow::Result<u32> {
+        let mut prover_storage = self
+            .prover_connection_pool
+            .connection()
+            .await
+            .context("failed to acquire DB connection for RecursionTipWitnessGenerator")?;
+        prover_storage
+            .fri_witness_generator_dal()
+            .get_recursion_tip_witness_job_attempts(*job_id)
+            .await
+            .map(|attempts| attempts.unwrap_or(0))
+            .context("failed to get job attempts for RecursionTipWitnessGenerator")
+    }
+}
+
+pub async fn prepare_job(
+    l1_batch_number: L1BatchNumber,
+    final_node_proof_job_ids: Vec<(u8, u32)>,
+    object_store: &dyn ObjectStore,
+) -> anyhow::Result<RecursionTipWitnessGeneratorJob> {
+    let started_at = Instant::now();
+    let recursion_tip_proofs =
+        load_proofs_for_recursion_tip(final_node_proof_job_ids, object_store).await?;
+    WITNESS_GENERATOR_METRICS.blob_fetch_time[&AggregationRound::RecursionTip.into()]
+        .observe(started_at.elapsed());
+
+    let keystore = Keystore::default();
+    let node_vk = keystore
+        .load_recursive_layer_verification_key(
+            ZkSyncRecursionLayerStorageType::NodeLayerCircuit as u8,
+        )
+        .context("get_recursive_layer_vk_for_circuit_type()")?;
+
+    let node_layer_vk_commitment = compute_node_vk_commitment(node_vk.clone());
+
+    let mut recursion_queues = vec![];
+    for circuit_id in BaseLayerCircuitType::as_iter_u8() {
+        let key = ClosedFormInputKey {
+            block_number: l1_batch_number,
+            circuit_id,
+        };
+        let ClosedFormInputWrapper(_, recursion_queue) = object_store.get(key).await.unwrap();
+        recursion_queues.push((circuit_id, recursion_queue));
+    }
+
+    assert!(
+        RECURSION_TIP_ARITY >= recursion_queues.len(),
+        "we have more circuits than supported"
+    );
+    let mut branch_circuit_type_set = [GoldilocksField::ZERO; RECURSION_TIP_ARITY];
+    let mut queue_set: [_; RECURSION_TIP_ARITY] =
+        std::array::from_fn(|_| QueueState::placeholder_witness());
+
+    for (index, (circuit_id, recursion_queue)) in recursion_queues.iter().enumerate() {
+        branch_circuit_type_set[index] = GoldilocksField::from_u64_unchecked(*circuit_id as u64);
+        queue_set[index] = take_sponge_like_queue_state_from_simulator(recursion_queue);
+    }
+
+    let leaf_vk_commits = get_leaf_vk_params(&keystore).context("get_leaf_vk_params()")?;
+    let leaf_layer_parameters: [RecursionLeafParametersWitness<GoldilocksField>; 16] =
```

Comment:
nit: consider adding the assert that leav_vk_commits.len() is 16.  - as otherwise this 'unwrap' is going to fail, and will be harder to debug.